### PR TITLE
[TF API] Implement initial TensorFlow dataset API with bug workarounds

### PR DESCRIPTION
**Pull request by rxwei**
## Introduction

TensorFlow's [data API](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) represents a potentially large set of elements and is to be used as an input pipeline for training. Transformation operators like `MapDataset`(https://github.com/tensorflow/tensorflow/blob/aed8f42bafabf11c5d92ce4109a5e0408b31f9c5/tensorflow/core/ops/dataset_ops.cc#L189) correspond exactly to Swift's `Sequence` methods. Iterator operator `IteratorGetNextAsOptional` corresponds exactly to Swift's `IteratorProtocol.next()`.

This patch adds `SingleValueDataset` type that implements `Sequence` and a `SingleValueDatasetIterator` type that implements `IteratorProtocol`. Both types use bare-form `#tfop` to wire up the dataset ops.

Kickstarts [SR-8438](https://bugs.swift.org/browse/SR-8438).

## Brief design rationale

Due to multiple language limitations, e.g. variadic generics and variadic tuples, we cannot have a single `Dataset` type that works for any nest-like structure. Using arrays or other dynamic data structures on the other hand would fail to fulfill efficient Graph Program Extraction. We start with this simple model and specialize dataset types for different common result types, e.g. `DoubleValueDataset`, `TripleValueDataset`, and `QuadrupleValueDataset`. Whenever the type system is good enough to generalize, we replace them with a type alias.

This API is going to unblock demos and tutorials instead of pursing full generality of nest-like APIs. I'll write a holistic design document on Swift dataset API design later down the road.

## Known issues

- [SR-8463](https://bugs.swift.org/browse/SR-8463) blocks any API that would result into indirect passing, so `SingleValueDataset` is intentionally specialized on `Float` and will be made generic once the bug is fixed.

- As documented in [SR-8478](https://bugs.swift.org/browse/SR-8478) `@convention(tensorflow)` parameters won't accept functions of the same type. Until this is fixed, we cannot test higher order operators.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

File: test/TensorFlowRuntime/dataset_api.swift
Comment (mhong) [inline test/TensorFlowRuntime/dataset_api.swift]: i'm not sure if the support of iterating over dataset could be a useful short-hand, or could lead to confusion.

should we stick to the TF dataset API design, where user must first create an iterator out of a dataset?

File: stdlib/public/TensorFlow/Dataset.swift
Comment (mhong) [inline stdlib/public/TensorFlow/Dataset.swift]: another design option is to avoid exposing the "cardinality" dimension of the dataset (e.g. SingleValueDataset, DoubleValueDataset etc) into different Dataset structs, and instead extend makeIterator() with param(s) to indicate the cardinality of the iterator we create.

This way when we later of datasets that read from different sources (files, etc), we won't need a corss-product of dataset structs from {single, double, tripple value ... } X {data set from tensor, data set from file, etc).

 WDYT?

**Review by mhong — COMMENTED** [review]
Nice progress!

File: stdlib/public/TensorFlow/Dataset.swift
Comment (dan-zheng) [reply stdlib/public/TensorFlow/Dataset.swift]: @mhong What do you mean exactly by "extending makeIterator() with param(s)"?
The `Sequence` protocols requires a [`makeIterator()`](https://developer.apple.com/documentation/swift/sequence/2885155-makeiterator) method that takes no arguments.

Eventually, I think the true way to generalize over `n` outputs is variadic generics.

Comment (mhong) [reply stdlib/public/TensorFlow/Dataset.swift]: Why do we need to conform dataset to `Sequence` protocol?

Comment (dan-zheng) [reply stdlib/public/TensorFlow/Dataset.swift]: `for-in` loops work on types that conform to `Sequence`.
This allows us to write things like `for item in dataset { ... }`, as you can see in the tests.

Comment (rxwei) [reply stdlib/public/TensorFlow/Dataset.swift]: I have explanation in the PR description. Please take a read. It is a type system limitation.

File: stdlib/public/TensorFlow/Dataset.swift
Comment (rxwei) [reply stdlib/public/TensorFlow/Dataset.swift]: As described in the PR description, higher order function tests are blocked by a compiler bug.

File: test/TensorFlowRuntime/dataset_api.swift
Comment (rxwei) [reply test/TensorFlowRuntime/dataset_api.swift]: You seem to suggest that TensorFlow does not support iterating over, but that’s not the case. TensorFlow’s data API *does* support being iterated over. Python’s tf.data.Dataset conforms to Python’s __iter__ protocol and *can* be iterated over. 

Plus, the entire dataset is consistent with standard Swift protocols both semantics-wise and API-wise. I don’t think I can see any arguments for confusion from your speculation.

File: stdlib/public/TensorFlow/Dataset.swift
Comment (rxwei) [reply stdlib/public/TensorFlow/Dataset.swift]: > extend makeIterator() with param(s) to indicate the cardinality of the iterator we create.

This is not a viable design. A Dataset and an Iterator have the same output nesting structure, like “cardinality” you referred to.

Comment (rxwei) [reply stdlib/public/TensorFlow/Dataset.swift]: > Eventually, I think the true way to generalize over `n` outputs is variadic generics. 

Dan is correct. Variadic generics are the only way to support it natively. It’s not possible in Swift today.

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

File: test/TensorFlowRuntime/dataset_api.swift
Comment (mhong) [reply test/TensorFlowRuntime/dataset_api.swift]: From https://www.tensorflow.org/guide/datasets, we have:

> Once you have built a Dataset to represent your input data, the next step is to create an Iterator to access elements from that dataset.

If we don't make dataset conform to sequence, the benefits are:

1. allow us to customize the makeIterator() API, and support things like make_one_shot_iterator and make_initializable_iterator if needed (see https://www.tensorflow.org/guide/datasets).

2. remove the need to define what it means to iterate over a dataset. is it a short hand for creating a one shot iterator? can i iterate over a dataset multiple times with a sequential list of for-in loops? can i create multiple threads that concurrently iterate over the same dataset object?

It seems simpler conceptually, if we define a dataset to be a stateless object -- a simple descriptor of how we read and transform the data  (e.g. from file, and apply shuffling, etc). To do the actual reading, create an iterator, which is stateful and not thread-safe. The use cases can then be:
-  if you want to iterate over the same dataset in multiple concurrent threads, where each thread reads its entire copy of data, create one iterator for each such thread. no synchronization is needed.
- if you want multiple threads to collaborate in reading from the same iterator, use a mutex to protect that iterator get_next calls. alternatively, have a single thread drive the get_next calls, and then dispatch/multiplex the output to multiple consumers.

File: stdlib/public/TensorFlow/Dataset.swift
Comment (mhong) [reply stdlib/public/TensorFlow/Dataset.swift]: Thanks. I didn't notice they are higher order functions. :)

File: test/TensorFlowRuntime/dataset_api.swift
Comment (rxwei) [reply test/TensorFlowRuntime/dataset_api.swift]: > 1. allow us to customize the makeIterator() API, and support things like make_one_shot_iterator and make_initializable_iterator if needed (see https://www.tensorflow.org/guide/datasets).

Yes, different iterators will be supported and I have considered this. This is not a benefit of not conforming to `Sequence`; this is a benefit of adding more APIs.

```swift
extension SingleValueDataset {
  func makeOneShotIterator() -> OneShotDatasetIterator<Scalar>
  func makeInitializableIterator() -> InitializableDatasetIterator<Scalar>
}
```

But a sequence like `tf.Dataset` has well-defined semantics for [what it means to iterator over elements a dataset](https://github.com/tensorflow/tensorflow/blob/25c197e02393bd44f50079945409009dd4d434f8/tensorflow/python/data/ops/dataset_ops.py#L113). That is why iterating over a dataset is very naturally expressed like iterating over a sequence in TensorFlow Eager. The user certainly has the ability to choose what iterators they use.

> 2. remove the need to define what it means to iterate over a dataset. is it a short hand for creating a one shot iterator?

It creates a default iterator for traversing a sequence, like Python's `__iter__` which internally calls `MakeIterator`. 

A one-shot iterator is the most flexible API because it allows the user to specify a generator. This API is going to exist and is orthogonal to whether `SingleValueDataset` conforms to `Sequence`. It can either be modeled as a separate type like Swift's [`AnyIterator`](https://developer.apple.com/documentation/swift/anysequence/2854368-init), or can just be created using a different `init` that uses `OneShotIterator` op: 

```swift
extension SingleValueDatasetIterator {
  init(datasetFactory: @convention(tensorflow) () -> SingleValueDataset<Scalar>)
}
```

The underlying mechanism is analogous to `datasetFactory().makeIterator()`, except that this makes the dataset closed over by the closure and thereby anonymous.

> can i iterate over a dataset multiple times with a sequential list of for-in loops? 

Yes, this is the fundamental semantics of TensorFlow's Dataset object. It is pure and can be iterated over multiple times.

> can i create multiple threads that concurrently iterate over the same dataset object?

By value semantics of Dataset, certainly yes. If you ask in a context of distributed programming, this is a also general question for TensorFlow's `Dataset` API. In both worlds, `Dataset` is stateless and `Iterator` is stateful.

> It seems simpler conceptually, if we define a dataset to be a stateless object -- a simple descriptor of how we read and transform the data (e.g. from file, and apply shuffling, etc). To do the actual reading, create an iterator, which is stateful and not thread-safe.

This *is* the current design. `Dataset` already is a stateless object by value semantics and it can be iterated over multiple times. The semantics is exactly the same as TensorFlow's `Dataset`.

You seem to believe that a `for`-`in` loop mutates the dataset, is that right? That's not the case because in Python or Swift because, a `for`-`in` loop creating an stateful iterator internally to iterate over. It is not mutating the dataset.

> * if you want to iterate over the same dataset in multiple concurrent threads, where each thread reads its entire copy of data, create one iterator for each such thread. no synchronization is needed.
> * if you want multiple threads to collaborate in reading from the same iterator, use a mutex to protect that iterator get_next calls. alternatively, have a single thread drive the get_next calls, and then dispatch/multiplex the output to multiple consumers.

Correct. With value semantics of `Dataset`, this is the exact behavior that this API offers. This is completely orthogonal to whether `Dataset` conforms to `Sequence`. Note that Swift's `Sequence` does not mean that the conforming object is stateful. It means that 1) an iterator can be created for that object and 2) there are functional operators the can be applied to the object to transform the sequence. This is exactly consistent with TensorFlow's `Dataset` and `Iterator` design. So I believe there's a misunderstanding here.

Comment (mhong) [reply test/TensorFlowRuntime/dataset_api.swift]: Ok, thanks for the various context. It seems we are agreeing on the same basic design, and the iteration support over a dataset is just syntax sugar that internally creates an iterator obj for each for-in loop. This is fine.

A few suggestions:
1. It'd be nice to not have the cardinality info leaked into on both dataset and iterator structs (returning a `OneShotDatasetIterator<Scalar>` is too much detail)
2. Can we model cardinality via an array of meta types instead of template args?

The rough design would look like:
```swift
protocol Dataset {
  // DatasetIterator is a protocol, so that user need not deal with concrete types of iterators.
  func makeOneShotIterator() -> DatasetIterator
  func makeInitializableIterator() -> DatasetIterator
  ...
  // if we want to conform to Sequence protocol, let makeIterator() map to one of the above make methods?
}

struct FileDataset : Dataset {
  // an init method over a file path
  public init(types: [AcceleratorByTensorFlow], shapes: [TensorShape], filePath: String) {
    ...
  }
  // implement the protocol methods from Dataset
  // e.g. makeOneShotIterator() creates an iterator based on the types and shapes known to the dataset.
}

struct TensorSliceDataset : Dataset {
  // an init method over a tensor
  // implement the protocol methods from Dataset
}
...
```

**Review by dan-zheng — APPROVED** [review]
LGTM. The dataset design brings together Swift's `Sequence` protocol and TensorFlow datasets and iterators in a natural way. I really like the example code in the tests and I'm excited to see further iteration on this design. Great work!

File: stdlib/public/TensorFlow/Dataset.swift
Comment (rxwei) [reply stdlib/public/TensorFlow/Dataset.swift]: Done

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
The CI issue has something to do with the TensorFlow version. Just updated that.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
Ok, it fails to clone https://github.com/tensorflow/swift-bindings for some reason.

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
Merging to unblock data API development. More comments welcome!

File: stdlib/public/TensorFlow/Dataset.swift
Comment (lattner) [inline stdlib/public/TensorFlow/Dataset.swift]: Is this the correct bug number?

Comment (rxwei) [reply stdlib/public/TensorFlow/Dataset.swift]: It is. When SingleValueDataset or SingleValueDatasetIterator is on the stack, const eval won’t work.

File: stdlib/public/TensorFlow/Dataset.swift
Comment (lattner) [inline stdlib/public/TensorFlow/Dataset.swift]: This isn't deabstraction's fault - this is XLA not supporting dynamic shapes.  Deabstraction can't magically know the shape of the dataset you've constructed with this API.

File: stdlib/public/TensorFlow/Dataset.swift
Comment (lattner) [reply stdlib/public/TensorFlow/Dataset.swift]: I don't understand the exchange here.  I think that Mingsheng has the right intuition - we shouldn't have one Swift type for each dataset element count that becomes interesting in the future.  We don't have the mechanics to support this at the moment, but we should build towards it.  I don't think we need variadic tuples, we just need to eventually have something like:

struct DataSet<T> : Sequence {
  typealias Element = T
...
}

Then you could use DataSet<Tensor<Float>> for a single value dataset.  If you want to support multiple value datasets, you'd define a struct like:

struct MyDataSetRecord {
  var x : Tensor<Float>
  var y : Tensor<Int8>
}

and use DataSet<MyDataSetRecord>.  I think this should compose together naturally, but we are missing the support for #tfop type inference to handle returns of structs like MyDataSetRecord.  If we handled them as multiple returns of the corresponding tensor elements, then it should all "just work" with the existing dataset #tfop's AFAIK.

File: test/TensorFlowRuntime/dataset_api.swift
Comment (lattner) [reply test/TensorFlowRuntime/dataset_api.swift]: I don't see a need for variadic generics here.  Why can't you use structs as suggested above, or use tuple types, e.g.:

```
typealias SingleValueDataset<T> = Dataset<T>
typealias DoubleValueDataset<T, U> = Dataset<(T, U)>
typealias TripleValueDataset<T, U, V> = Dataset<(T, U, V)>
typealias QuadrupleValueDataset<T, U, V, W> = Dataset<(T, U, V, W)>
```

The only downside to using tuples like I showed above is that tuples can't conform to protocols, so you can't constrain the element type of DataSet.  That said, structs should work fine and solve that problem.

File: stdlib/public/TensorFlow/Dataset.swift
Comment (rxwei) [reply stdlib/public/TensorFlow/Dataset.swift]: > I don't understand the exchange here. I think that Mingsheng has the right intuition - we shouldn't have one Swift type for each dataset element count that becomes interesting in the future.

Using the single type instead of multiple types was my first consideration. I agree with your intuition as well, and I also dislike using specialized types for datasets with different result values. However, it won't work in today's `#tfop` because one can't turn an aggregate into a list of metatypes of tensor dtypes to pass in dataset ops.

Moreover, it is questionable whether `#tfop` can be extended to support dataset op requirements. Element shapes and element dtypes are required by dataset ops. How would you store element shapes such that they can be passed as an array? And how would you do so for scalar metatypes? 

In your `MyDataSetRecord` case, we'd need to
1) constrain all elements to `Tensor` (this could be done via protocol conformance and derived conformances, but it's very complex for this purpose),

2) extract the metatype from each field's associated `Scalar` type and pass them as a const array to dataset tfops (this is impossible today and it would complicate `#tfop` a lot), and

3) either have a struct parallel to `MyDataSetRecord` to store all element shapes, or store an element shape as part of the `MyDataSetRecord`.
    ```swift
    struct MyDataSetRecord {
      var x: (Tensor<Float>, TensorShape)
      var y: (Tensor<Int8>, TensorShape)
    }
    ```
    This would again make `#tfop` very complicated for extracting struct members. 

> If we handled them as multiple returns of the corresponding tensor elements, then it should all "just work" with the existing dataset #tfop's AFAIK.

As shown above, it is not just extending `#tfop` to support multiple returns, but rather extending `#tfop` to support a form of operations analogous to nest mapping. And that seems very far away from today's `#tfop` design, and the complexity is unmeasured.

File: stdlib/public/TensorFlow/Dataset.swift
Comment (rxwei) [reply stdlib/public/TensorFlow/Dataset.swift]: I'm aware that deabstraction cannot handle shapes today. Maybe I did not use the best word to describe this. TensorFlow has shape inference C APIs, and you can call that to get the shape of some graph node, if not specified by user code. This idea was brought up by Anthony a few days ago - maybe it's worth considering delaying constant checking of shape attributes to graph lowering so that TF APIs can help.

File: test/TensorFlowRuntime/dataset_api.swift
Comment (rxwei) [reply test/TensorFlowRuntime/dataset_api.swift]: > The only downside to using tuples like I showed above is that tuples can't conform to protocols.

That's what I meant by "variadic tuples" and I should've made that clearer. I believe the consensus on Swift evolution was that variadic tuples and tuples being an alias to `Tuple<T...>` are a prerequisite for making tuples able to conform to protocols.

I showed some problems with your struct approach to handle multiple results in a separate comment. I would appreciate ideas on the tactics to resolve this. Again, I'm also not a fan of having multiple types representing datasets with different arity, but the technical constraint with `#tfop` is a real bottleneck.

**Pull request merged**
